relay-metrics/src/aggregation.rs

```diff
@@ -1357,6 +1324,160 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
+/// - If flushing fails and should be retried at a later time, respond with `Err` containing the
+///   failed buckets. They will be merged back into the aggregator and flushed at a later time.
+#[derive(Clone, Debug)]
+pub struct FlushBuckets {
+    /// The project key.
+    pub project_key: ProjectKey,
+    /// The logical partition to send this batch to.
+    pub partition_key: Option<u64>,
+    /// The buckets to be flushed.
+    pub buckets: Vec<Bucket>,
+}
+
+// NOTE(): required by ProjectCache and will be removed with ProjectCache migration.
+impl Message for FlushBuckets {
+    type Result = ();
+}
+
+/// A message containing a list of [`Metric`]s to be inserted into the aggregator.
+#[derive(Debug)]
+pub struct InsertMetrics {
+    project_key: ProjectKey,
+    metrics: Vec<Metric>,
+}
+
+// NOTE: required by ProjectCache and will be removed with ProjectCache migration.
+impl Message for InsertMetrics {
+    type Result = Result<(), AggregateMetricsError>;
+}
+
+impl InsertMetrics {
+    /// Creates a new message containing a list of [`Metric`]s.
+    pub fn new<I>(project_key: ProjectKey, metrics: I) -> Self
+    where
+        I: IntoIterator<Item = Metric>,
+    {
+        Self {
+            project_key,
+            metrics: metrics.into_iter().collect(),
+        }
+    }
+
+    /// Returns the `ProjectKey` for the the current `InsertMetrics` message.
+    pub fn project_key(&self) -> ProjectKey {
+        self.project_key
+    }
+
+    /// Returns the list of the metrics in the current `InsertMetrics` message, consuming the
+    /// message itself.
+    pub fn metrics(self) -> Vec<Metric> {
+        self.metrics
+    }
+}
+
+/// A message containing a list of [`Bucket`]s to be inserted into the aggregator.
+#[derive(Debug)]
+pub struct MergeBuckets {
+    project_key: ProjectKey,
+    buckets: Vec<Bucket>,
+}
+
+// NOTE: required by ProjectCache and will be removed with ProjectCache migration.
+impl Message for MergeBuckets {
+    type Result = Result<(), AggregateMetricsError>;
+}
+
+impl MergeBuckets {
+    /// Creates a new message containing a list of [`Bucket`]s.
+    pub fn new(project_key: ProjectKey, buckets: Vec<Bucket>) -> Self {
+        Self {
+            project_key,
+            buckets,
+        }
+    }
+
+    /// Returns the `ProjectKey` for the the current `MergeBuckets` message.
+    pub fn project_key(&self) -> ProjectKey {
+        self.project_key
+    }
+
+    /// Returns the list of the buckets in the current `MergeBuckets` message, consuming the
+    /// message itself.
+    pub fn buckets(self) -> Vec<Bucket> {
+        self.buckets
+    }
+}
+
+/// Aggregator service interface
+#[derive(Debug)]
+pub enum Aggregator {
+    /// The health check message which makes sure that the service can accept the requests now.
+    AcceptsMetrics(AcceptsMetrics, Sender<bool>),
+    /// Insert metrics.
+    InsertMetrics(InsertMetrics),
+    /// Merge the buckets.
+    MergeBuckets(MergeBuckets),
```

Comment:
The actix `Message` implementations return a `Result` with an aggregation error on failure. For the time being, we should not change that contract and have a `Sender<Result<(), AggregateMetricsError>>` for these two.

Reply:
As discussed, we never checked for the returned errors for these messages, so to simplify our interface we agreed on keeping `NoResponse`.